src/third_party/wiredtiger/src/meta/meta_ckpt.c

```diff
@@ -240,6 +243,246 @@ __ckpt_compare_order(const void *a, const void *b)
+	 * searching from the end position of the last attribute. However,
+	 * we are robust against out of order attributes by resetting "s"
+	 * to the beginning of the string to search again if an attribute
+	 * is not parsed in its expected location.
+	 */
+	if (__ckpt_parse_attribute("order=", &s, val, &len)) {
+		s = str;
+		__ckpt_parse_attribute("order=", &s, val, &len);
+		/* We have failed to find the attribute. */
+		if (len == 0) {
+			goto format;
+		}
+	}
+
+	ckpt->order = strtol(val, NULL, 10);
+
+	if (__ckpt_parse_attribute("time=", &s, val, &len)) {
+		s = str;
+		__ckpt_parse_attribute("time=", &s, val, &len);
+		if (len == 0) {
+			goto format;
+		}
+	}
+
+	ckpt->sec = strtoul(val, NULL, 10);
+
+	if (__ckpt_parse_attribute("size=", &s, val, &len)) {
+		s = str;
+		__ckpt_parse_attribute("size=", &s, val, &len);
+		if (len == 0) {
+			goto format;
+		}
+	}
+
+	ckpt->ckpt_size = strtoul(val, NULL, 10);
+
+	if (__ckpt_parse_attribute("write_gen=", &s, val, &len)) {
+		s = str;
+		__ckpt_parse_attribute("write_gen=", &s, val, &len);
+		if (len == 0) {
+			goto format;
+		}
+	}
+
+	ckpt->write_gen = strtoul(val, NULL, 10);
+
+	return (0);
+
+format:
+	WT_RET_MSG(session, WT_ERROR, "corrupted checkpoint list");
+}
+
+/*
+ * __wt_meta_ckptlist_get_custom --
+ *	Load all available checkpoint information for a file.
+ *  This is an alterative to __wt_meta_ckptlist_get implemented with
+ *  a faster parser.
+ */
+int
+__wt_meta_ckptlist_get_custom(
+    WT_SESSION_IMPL *session, const char *fname, WT_CKPT **ckptbasep)
+{
+	WT_CKPT *ckpt, *ckptbase;
+	WT_DECL_ITEM(buf);
+	WT_DECL_RET;
+	size_t allocated, slot;
+	char *config;
+	char *found = NULL;
+	char *name_end = NULL;
+	uint64_t time_start_1, time_stop_1;
+	uint64_t time_start_2, time_stop_2;
+
+	*ckptbasep = NULL;
+
+	ckptbase = NULL;
+	allocated = slot = 0;
+	config = NULL;
+
+	time_start_1 = __wt_clock(session);
+
+	/* Retrieve the metadata information for the file. */
+	WT_RET(__wt_metadata_search(session, fname, &config));
+
+	time_stop_1 = __wt_clock(session);
+	WT_STAT_CONN_INCRV(session, txn_checkpoint_meta_ckptlist_search,
+		WT_CLOCKDIFF_US(time_stop_1, time_start_1));
+
+	/* Load any existing checkpoints into the array. */
+	WT_ERR(__wt_scr_alloc(session, 0, &buf));
+
+	found = config;
+
+	time_start_2 = __wt_clock(session);
+
+	/*
+	 * Find all instances of WiredTigerCheckpoint metadata and
+	 * place each into a slot.
+	 */
+	while ((found = strstr(found, "WiredTigerCheckpoint."))) {
+		WT_ERR(__wt_realloc_def(
+			session, &allocated, slot + 1, &ckptbase));
+		ckpt = &ckptbase[slot];
+
+		/*
+		 * Length of "WiredTigerCheckpoint." is 21, start searching
+		 * for the value there.
+		 */
+		name_end = strchr(found + 21, '=');
+
+		WT_RET(__wt_strndup(session, found,
+			(int)(name_end - found), &ckpt->name));
+
+		/*
+		 * Begin parsing attributes starting from the end of the
+		 * name string.
+		 */
+		__ckpt_parse_attributes(session, name_end, ckpt);
```

Comment:
We could probably start at name_end + 1.

Reply:
Sure